File: DSA/Anisha_Parveen/Day_9_Ques_1.cpp
#include <bits/stdc++.h>
using namespace std;

struct MaxStack {
    stack<int> mainStack;
    stack<int> maxStack;
};

void push(MaxStack& stack, int x) {
    stack.mainStack.push(x);
    if (stack.maxStack.empty() || x >= stack.maxStack.top()) {
        stack.maxStack.push(x);
    } else {
        stack.maxStack.push(stack.maxStack.top());
    }
}

void pop(MaxStack& stack) {
    if (stack.mainStack.empty()) {
        cerr << "Stack is empty" << endl;
        return;
    }
    stack.mainStack.pop();
    stack.maxStack.pop();
}

int getMax(const MaxStack& stack) {
    if (stack.maxStack.empty()) {
        cerr << "Stack is empty" << endl;
        return -1;
    }
    return stack.maxStack.top();
}

int main() {
    MaxStack stack;
    push(stack, 1);
    push(stack, 2);
    cout << getMax(stack) << endl; 
    pop(stack);
    cout << getMax(stack) << endl; 
    return 0;
}

File: DSA/Asmi_Garg/day9_ques1.java
import java.util.Stack;

public class MaxStack {
    private Stack<Integer> stack;
    private Stack<Integer> maxStack;

    public MaxStack() {
        stack = new Stack<>();
        maxStack = new Stack<>();
    }

    public void push(int x) {
        stack.push(x);
        if (maxStack.isEmpty() || x >= maxStack.peek()) {
            maxStack.push(x);
        }
    }

    public void pop() {
        if (!stack.isEmpty()) {
            int topElement = stack.pop();
            if (topElement == maxStack.peek()) {
                maxStack.pop();
            }
        }
    }

    public int getMax() {
        if (!maxStack.isEmpty()) {
            return maxStack.peek();
        }
        throw new RuntimeException("Stack is empty");
    }

    public static void main(String[] args) {
        MaxStack stack = new MaxStack();
        stack.push(1);
        stack.push(2);
        System.out.println(stack.getMax()); // Output: 2
        stack.pop();
        System.out.println(stack.getMax()); // Output: 1
    }
}
